lib/flight_job/commands/run_monitor.rb

```diff
@@ -29,13 +29,8 @@ module FlightJob
   module Commands
     class RunMonitor < Command
       def run
-        Job.load_active.each do |job|
-          if job.valid?(:monitor)
-            job.monitor
-          else
-            job.transition_inactive
-          end
-        end
+        Job.transition_inactive
+        Job.load_all
```

Comment:
The `run_monitor` command has two purposes:

1. Pick up jobs where `submit.sh` failed (aka crashed an burned), and
2. Run the monitor script for all non-terminal jobs.

---

## Use case 1: Transition inactive

Use case one is fairly easy, glob for all the `active` files and check the scheduler if it knows anything about it. It then runs some logic to determine if the job was submitted to the scheduler or if it failed.

Regardless, the initial_metadata_path _should_ be deleted after `Job#transition_inactive` is called.

*NOTE:* `active_path` has been renamed as it is rather obscure in meaning. `initial_metadata_path` is far more descriptive. The are still some methods which use the `active`/`inactive` terminology.

---

## Use case 2: Update Non-Terminal Jobs

The secondary use case is to update the metadata state for non-terminal jobs. This is a bit trickier as the state is stored within the `metadata`. Currently the only way to check this `state` is to load all the `jobs`. This will trigger the `monitor.sh` to be ran (when required).

This is an O(N) operation, and could be fairly IO intensive.

---

## Proposed fix: Delay deleting initial_metadata_path

Currently the existence of the `initial_metadata_path` is used to denote:
_Did the job get submitted to the scheduler correctly?_

Instead its meaning could be updated to:
_Is the job in a terminal state?_

This way `Job.transition_inactive` would handle both the use cases mentioned above.

---

## Additional Fix: Migration Config File

The above fix is not currently possible as changing the meaning of `initial_metadata_path` is not backwards compatible. Any previously submitted jobs in a `pending`/`running` state would have already deleted the file.

A possible way around this is to add a new "config_file":
`~/.local/flight/job/.migration.yaml`

This would save the last known version of `flight-job` used. The next time `flight-job` is ran, it will attempt to do the following:

* Check if the config exists and is at least the current version,
* Do nothing if it is at-least the current version,
* Otherwise, loop through all the `jobs` to added the `initial_metadata_path` back (if required), and
* Update and save the migration file.

*NOTE*: Ideally the logic for the migration should not depend on the `models`. Instead the migration should be attempted syntactically. This is to prevent the migration becoming coupled to a particularly version of `Job`. This may or may not be possible.

Reply:
What I'm taking from that is that `Job.load_all` calls `Job#monitor` as a side effect. I suspect that we'll likely want to decouple that at some point.  But it should be good enough for now.

Reply:
Ah okay

I could add a `monitor_all` method if you like? But currently `load_all` and `monitor_all` is the same thing. It isn't possible to monitor a job without loading the metadata